### PR TITLE
Bugfix: Webserver returns 500 for POST requests to api/dag/*/dagrun from anonymous user

### LINKED ISSUE #36110: Webserver returns 500 for POST requests to api/dag/*/dagrun from anonymous user
### Apache Airflow version

2.7.3

### What happened

After upgrade to 2.7.3 I noticed that webserver returns 500 for POST requests with a `note` field. After some investigation it seems that None in [user_id](https://github.com/apache/airflow/blob/main/airflow/models/dagrun.py#L1524) converted to str causing it: 
```
[SQL: INSERT INTO dag_run_note (user_id, dag_run_id, content, created_at, updated_at) VALUES (%(user_id)s, %(dag_run_id)s, %(content)s, %(created_at)s, %(updated_at)s)]
[parameters: {'user_id': 'None', 'dag_run_id': 2944, 'content': 'test', 'created_at': datetime.datetime(2023, 12, 7, 17, 42, 15, 390085, tzinfo=Timezone('UTC')), 'updated_at': datetime.datetime(2023, 12, 7, 17, 42, 15, 390096, tzinfo=Timezone('UTC'))}]
rundag [07/Dec/2023:17:42:15 +0000] "POST /api/v1/dags/*/dagRuns HTTP/1.1" 500 1539 "-" "python-requests/2.31.0"
```
That's probably not an issue of Airflow itself, but SQLAlchemy 1.4. it's probably fixed in 2.0.

### What you think should happen instead

Dagrun should be successfully registered by airflow and note should be saved into `dag_run_note` table.

### How to reproduce
1. Disable Auth to allow unauthorized access to the API in `webserver_config.py`
```
AUTH_ROLE_PUBLIC = "Admin"
```
2. Try to schedule a run via `DAGRun` API
```python
import requests
dag_run_url="https://airflow/api/v1/dags/test/dagRuns"
r = requests.post(
    url=dag_run_url, json={"dag_run_id": "test_me", "note":"note", "conf": {}}
)
assert r.status_code == 500

print(r.text)

```

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Other Docker-based deployment

### Deployment details

Docker image: apache/airflow:2.7.3-python3.9
DB: Postgresql

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by pateash**
closes: #36110
related: #36206

---

## Description
While creating a dag_note in anonymous mode ( API auth set to  backend default ) 

`
export AIRFLOW__API__AUTH_BACKENDS= airflow.api.auth.backend.default
`
user_id is coming as 'None' as userContext is not available.
which is invalid for an integer field in the database and causing SQL alchemy Exception.

## Proposed Solution
Check if the user_id is None, we will not cast it into string type which will resolve the issue.

### CODE REVIEW

**Review by ephraimbuddy — DISMISSED** [review]
You can only disable authentication for the experimental REST API not the stable REST API so this is not an issue. See: https://airflow.apache.org/docs/apache-airflow/stable/security/api.html#disable-authentication

**Review by hussein-awala — COMMENTED** [review]
Could you add a test for this change?

**Comment by pateash** [comment]
> Could you add a test for this change?

I am trying, but not able to create a client without any auth.


> Could you add a test for this change?
@hussein-awala 
I have added a test, which is utilizing **Anonymous** user, but still getting 401 (Unauthorized),
might need some help in getting this correct.

File: tests/api_connexion/endpoints/test_dag_run_endpoint.py [hunk 1]
```diff
@@ -1861,3 +1861,17 @@ def test_should_respond_404(self):
...
+
+        dag_runs = self._create_test_dag_run(DagRunState.SUCCESS)
+        session.add_all(dag_runs)
+        session.commit()
+        created_dr = dag_runs[0]
+        response = self.client.post(
+            f"api/v1/dags/{created_dr.dag_id}/dagRuns",
+            json={"dag_run_id": "TEST_DAG_RUN_ID_1", "note": "I am setting a note with anonymous user"},
+        )
+        assert response.status_code == 200
```
Comment (pateash) [inline tests/api_connexion/endpoints/test_dag_run_endpoint.py:1883]: still getting 401, unauthorized 🤔 
 
image

Comment (hussein-awala) [reply tests/api_connexion/endpoints/test_dag_run_endpoint.py:1883]: There were some problems in the implemented test (wrong endpoint, wrong method, the application was already created and configured, etc.). I just pushed a commit to fix it, and I tested it before (I had SQL integrity exception) and after your change (it works).

**Review by hussein-awala — APPROVED** [review]
LGTM

File: airflow/auth/managers/base_auth_manager.py [hunk 2]
```diff
@@ -99,13 +99,13 @@ def get_user_display_name(self) -> str:
...
 
-    def get_user_id(self) -> str:
+    def get_user_id(self) -> str | None:
         """Return the user ID associated to the user in session."""
         user = self.get_user()
         if not user:
             self.log.error("Calling 'get_user_id()' but the user is not signed in.")
             raise AirflowException("The user must be signed in.")
-        return str(user.get_id())
+        return str(user.get_id()) if user.get_id() else None
```
Comment (uranusjr) [inline airflow/auth/managers/base_auth_manager.py]:
```suggestion
        if (user_id := user.get_id()):
            return str(user_id)
        return None
```

Depending on the auth backend, the function may not be cheap.

**Review by uranusjr — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Comment by pateash** [comment]
one of the test started failing after the rebase ( not related to this changes ) 
image

**Comment by potiuk** [comment]
Rebased. Likely some intermittent issue or smth connected with FAB changes being merged.

**Review by Lee-W — APPROVED** [review]

**Comment by potiuk** [comment]
Nope. Looks like real test failure that you need to fix - likely coming from recent changes in FAB structure.

**Comment by hussein-awala** [comment]
I think the change I pushed is not isolated, and it impacts the application context for other tests, I will take a second look this evening.

**Comment by hussein-awala** [comment]
I fixed it by updating the configuration after initializing the flask application: https://github.com/apache/airflow/pull/36275/commits/826f4e94d957366b7e7322d3e64c35d89dfaa0bb

**Pull request merged**